### PR TITLE
Dt 1377 document filter

### PR DESCRIPTION
**Pull request by andymarke**
Add the ability to filter by document category and for one category also by type

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/delius/controller/secure/DocumentResource.java [hunk 1]
```diff
@@ -34,25 +36,65 @@
 @AllArgsConstructor
 @PreAuthorize("hasRole('ROLE_COMMUNITY')")
 public class DocumentResource {
+    private static final String categoryDocumentation = "" +
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/delius/controller/secure/DocumentResource.java:39]: See if there is a clever way to use Enum

File: src/main/java/uk/gov/justice/digital/delius/service/DocumentService.java [hunk 2]
```diff
@@ -341,9 +345,10 @@ private Long eventId(ContactDocument document) {
                 .orElseGet(ImmutableList::of);
     }
 
-    private List<OffenderDocumentDetail> previousConvictions(Offender offender) {
+    private List<OffenderDocumentDetail> previousConvictions(Offender offender, Predicate<Offender> previousConvictionCheck) {
         return Optional.of(offender)
                 .filter(this::hasPreviousConvictions)
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/delius/service/DocumentService.java:350]: remove this line

File: src/main/java/uk/gov/justice/digital/delius/data/filters/DocumentFilter.java [hunk 3]
```diff
@@ -1,8 +1,100 @@
...
+
+    public static Either<String, DocumentFilter> of(String category, String type) {
+        if (StringUtils.isBlank(category) && StringUtils.isBlank(type)) {
+            return Either.right(NO_FILTER);
+        }
+        if (StringUtils.isBlank(category) && StringUtils.isNotBlank(type)) {
+            return Either.left(String.format("Type of %s was supplied but no category. Type can only be supplied when a valida category is supplied", type));
+        }
+
+        final var maybeCategory = Arrays.stream(Type.values()).filter(v -> v.name().equals(category)).findFirst();
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/delius/data/filters/DocumentFilter.java:48]: extract method

File: src/test/java/uk/gov/justice/digital/delius/service/DocumentServiceTest.java [hunk 4]
```diff
@@ -464,4 +480,128 @@ public void contactDocumentsDistributedToEachConvictionAndOffender() {
...
+            @Test
+            @DisplayName("Will only return documents for the category")
+            void willOnlyReturnDocumentsForTheCategory() {
+                final var documents = documentService.offenderDocumentsFor(1L, DocumentFilter
+                    .of("COURT_REPORT_DOCUMENT", null)
+                    .get());
+                assertThat(documents.getConvictions()).hasSize(1);
+                assertThat(documents.getConvictions().get(0).getDocuments()).hasSize(1);
+                assertThat(documents.getDocuments()).hasSize(0);
+            }
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/digital/delius/service/DocumentServiceTest.java:604]: Paramatised test

File: src/testIntegration/java/uk/gov/justice/digital/delius/controller/api/CourtReportAPITest.java [hunk 5]
```diff
@@ -55,7 +55,7 @@ public void canGetAllReportsForOffenderByCrn() {
             .body()
             .as(CourtReport[].class);
 
-        assertThat(courtReports).hasSize(1);
+        assertThat(courtReports).hasSize(2);
```
Comment (andymarke) [inline src/testIntegration/java/uk/gov/justice/digital/delius/controller/api/CourtReportAPITest.java:58]: just assert i have some documents

File: src/testIntegration/java/uk/gov/justice/digital/delius/controller/secure/DocumentResourceAPITest.java [hunk 6]
```diff
@@ -148,5 +256,101 @@ public void givenKnownCrnThenReturnDocuments() {
...
+                    .contentType(APPLICATION_JSON_VALUE)
+                    .when()
+                    .get("/offenders/crn/{crn}/documents/grouped", "X320741")
+                    .then()
+                    .statusCode(HttpStatus.OK.value())
+                    .extract()
+                    .body()
+                    .as(OffenderDocuments.class);
+
+                assertThat(countAllDocuments(offenderDocuments)).isEqualTo(24);
```
Comment (andymarke) [inline src/testIntegration/java/uk/gov/justice/digital/delius/controller/secure/DocumentResourceAPITest.java:278]: move constants

File: src/main/java/uk/gov/justice/digital/delius/data/filters/DocumentFilter.java [hunk 7]
```diff
@@ -1,8 +1,112 @@
...
     public static DocumentFilter noFilter() {
         return NO_FILTER;
     }
+
+    public static Either<String, DocumentFilter> of(String category, String type) {
+        if (StringUtils.isBlank(category) && StringUtils.isBlank(type)) {
+            return Either.right(NO_FILTER);
+        }
+        if (StringUtils.isBlank(category) && StringUtils.isNotBlank(type)) {
+            return Either.left(String.format("Type of %s was supplied but no category. Type can only be supplied when a valida category is supplied", type));
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/digital/delius/data/filters/DocumentFilter.java:47]: typo on `valida`

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**